lib/rules/no-constant-condition.js

```diff
@@ -66,15 +63,23 @@ module.exports = {
         function isLogicalIdentity(node, operator) {
             switch (node.type) {
                 case "Literal":
-                    return (operator === "||" && node.value === true) ||
-                           (operator === "&&" && node.value === false);
+                    return (operator === "||" && !!node.value) ||
+                           (operator === "&&" && !node.value);
```

Comment:
ESTree doesn't guarantee `node.value` is a valid value. For example, if you run ESLint on Node.js that doesn't support BigInt, `node.value` is `null` at bigint literals. It will cause false positive/negative. Similarly, RegExp literals that contain new syntax, bigdecimal literals (stage 1) are, too.

Reply:
How about reporting an unsupported error like the following code?

```js


class UnsupportedNodeError extends Error {}

function isNodeTruthy(node) {
    if (node.value === null && node.raw !== 'null') {
        throw new UnsupportedNodeError(`unsupported node: ${node.type}`)
    }

    return !!node.value;
}

function isLogicalIdentity(node, operator) {
    switch (node.type) {
        case "Literal":
            return (operator === "||" && isNodeTruthy(node)) ||
                (operator === "&&" && !isNodeTruthy(node));
        // ...
    }
}

function reportIfConstant(node) {
    try {
        if (node.test && isConstant(node.test, true)) {
            context.report({ node: node.test, messageId: "unexpected" });
        }
    } catch (error) {
        if (error instanceof UnsupportedNodeError) {
            context.report({ node: node.test, message: error.message })
        } else {
            throw error
        }
    }
}
```

Reply:
Even if the runtime doesn't support a syntax, ESLint should check the syntax successfully if ESLint supports it. In other words, the result of linting should be the same on all supported runtimes.

Reply:
>In other words, the result of linting should be the same on all supported runtimes.

Can we achieve runtime consistency by not supporting bigint on advanced runtime? Though bigint will not behave the same as numbers, just like numbers not behaving the same as booleans currently.

```js
function isNodeTruthy(node) {
    // TODO: bigint is not supported. wait till node v10.4.0+ is popular.
    if (typeof node.value === 'bigint' || (node.value === null && node.raw !== 'null')) {
        return false;
    }

    return !!node.value;
}
```

Reply:
As per ESTree spec, the `Literal` nodes of BigInt have `bigint` property with the text representation of the bigint value. And the `Literal` nodes of RegExp have `regex` property with `{ pattern: string; flags: string }`. The `value` property of `Literal` nodes of those may be `null`, so we have to check the `bigint`/`regex` properties to distinguish `null` literal and the others.

Reply:
Currently, eslint is using espree 6.2.1, which does not support bigint or regex with `u` flag. 🤦

Reply:
The actual version is using espree 7.1.0

https://github.com/eslint/eslint/blob/b7d79b12eb55f220f86367d52f9f5a2f503fd4c3/package.json#L59

Reply:
I was using an old version. Anyway, fallback handling is added. Your suggestions would be helpful.

https://github.com/eslint/eslint/pull/13245/files#diff-aee426385c16cd9eadb5fabd23dfa0bfR64-R84

Reply:
The code below is working on node 10.0.0.

```js
if(a);
if(/foo/ui);
if(0n);
if(1n && a);
if(1n || a);
```

```
  2:4  error  Unexpected constant condition  no-constant-condition
  3:4  error  Unexpected constant condition  no-constant-condition
  5:4  error  Unexpected constant condition  no-constant-condition
```